feat: add configure function for log level visibility

bgwdotdev opened the pull request.

bgwdotdev committed:
feat: add configure function for log level visibility

lpil reviewed (commented):
Hello! Could you add use of this function to the test please

bgwdotdev commented:
No bother! added to the test case and also updated the changelog if that's helpful too ^^;

cheers :)

bgwdotdev committed:
test: add `set_level` case/example

bgwdotdev committed:
docs: add `set_level` function docs

bgwdotdev commented:
just as a quick note, I added a little doc for the set_level function in-line with the docs you've added and rebased against main 

Additionally, I was going to update the return type to better return as per erlang docs https://www.erlang.org/doc/man/logger#set_primary_config-2 but I wasn't actually sure how to handle the `term()` type on the error since it's an any type? is there a reference example of how to handle this at all somewhere ^^?

lpil commented:
Let's use `Dynamic` for `term()` as we don't know what it is. Could be anything, and we don't care about it!

Thank you

bgwdotdev committed:
fix: change config return type to result

bgwdotdev commented:
ah! of course, forgot about dynamic. Now added that.

Wonder if it would make sense to also add helper functions for log levels to shorthand easier, for example

```gleam
logging.log(logging.Error, "my error")

logging.error("my error")
```

happy to throw in a function for each.

Was also considering if having a formatter for structured logging would make sense? pulling in gleam_json and maybe doing something like

```json

{ "level": "error", "date": 1715380045, "message": "my error" }

```
or something to that effect ^^?

Just some ideas if they were of interest, happy to implement.

bgwdotdev committed:
fix: add return type wrapper to allow casting into a result

lpil commented on src/logging.gleam:
```diff
@@ -33,3 +35,19 @@ pub fn log(level: LogLevel, message: String) -> Nil {
+
+/// Change the log visibility level to be output from the default of `Info`.
+///
+pub fn set_level(level: LogLevel) -> Result(Nil, dynamic.Dynamic) {
```
Please revert this to the previous API. I don't want to expose the error value at all, rather silently do nothing in that case like the existing code does.

bgwdotdev replied:
Oh ok! fair enough, will do!

lpil reviewed (commented):
> Wonder if it would make sense to also add helper functions for log levels to shorthand easier, for example
>
> Was also considering if having a formatter for structured logging would make sense? pulling in gleam_json and maybe doing something like

This package does nothing but offer some simple defaults for the Erlang logger. There already logging libraries that do these, best to use those instead.

bgwdotdev committed:
revert: remove return types from config change

bgwdotdev commented:
reverted! hopefully good to go now :) thanks for your feedback and time

lpil commented on src/logging.gleam:
```diff
@@ -33,3 +33,17 @@ pub fn log(level: LogLevel, message: String) -> Nil {
+@external(erlang, "logger", "set_primary_config")
+fn set_primary_config_level(key: Key, level: LogLevel) -> Nil
+
+/// Change the log visibility level to be output from the default of `Info`.
```
`Info` isn't the default for the Erlang logger, so let's remove that bit.

lpil commented on src/logging.gleam:
```diff
@@ -33,3 +33,17 @@ pub fn log(level: LogLevel, message: String) -> Nil {
 fn erlang_log(level: LogLevel, message: String) -> DoNotLeak
+
+@external(erlang, "logger", "set_primary_config")
+fn set_primary_config_level(key: Key, level: LogLevel) -> Nil
```
```suggestion
fn set_primary_config_level(key: Key, level: LogLevel) -> DoNotLeak
```

bgwdotdev replied:
I'm curious on this bit, is this just a type purely set for developer documentation purposes? not doing any special stuff in the ffi or anything?

lpil replied:
It's a private type defined above to ensure that any unknown value doesn't leak in to the public interface. The previous typing was incorrect and due to being public could leak due to a programmer mistake, which could potentially cause runtime issues.

lpil reviewed (approved):
Thank you 💜

bgwdotdev committed:
refactor: docs and internal output type

lpil reviewed (approved):
Thank you!!
